### PR TITLE
Fixes #18518 - Validate existence of docker image

### PR DESCRIPTION
**Pull request by bastilian**
Creating a container with an image that does not exist will cause the host creation to fail.

This validates the existence of the image and gives proper error feedback to the user in case it is missing:

screen shot 2017-02-16 at 15 03 07

### CODE REVIEW

**Comment by dLobatog** [comment]
@bastilian Thanks! I'm testing it now - it looks like it's missing a test as well 😄

**Comment by bastilian** [comment]
Adding some tests asap.

**Comment by bastilian** [comment]
Functional tests are in now. should the JS part be covered with integration tests as well?

The wording of the error is changed too, but only mentioning the registry name or "Docker Hub" and leaving out the compute resource.

And the view is now cleaned up and uses helpers as suggested.

**Comment by dLobatog** [comment]
Funny thing I found while testing this PR (tried to search for this thinking there would be no image like that, wrong):

screenshot from 2017-02-17 08-48-12

File: app/models/docker_container_wizard_states/image.rb [hunk 1]
```diff
@@ -4,8 +4,32 @@ class Image < ActiveRecord::Base
...
+      @registry_name ||= if registry_id
+        DockerRegistry.find(registry_id).fetch(:name, _('Unnamed Registry'))
+      else
+        'Docker Hub'
+      end
+    end
+
+    def validate_image_exists
+      unless compute_resource.exist? name
+        error_msg = _("Container image %{image_name} could not be found on %{registry}.") % {
```
Comment (dLobatog) [inline app/models/docker_container_wizard_states/image.rb]: When does this error message show up? I'm unable to see this error. Right now if I input a image that doesn't exist, and click on next, I get:

screenshot from 2017-02-17 08-53-12

If I try to write , I see this error (which is right, the image is not found in the compute resource):

screenshot from 2017-02-17 08-55-22

The error the issue describes should be displayed when the image is not available in the registry or the Docker host does not have it (e.g: if I had manually downloaded a 'fakenews:faketag' image to the Docker host, I should be able to move forward despite the image not being available, because it's on the Docker host)

**Comment by dLobatog** [comment]
@bastilian There are a few options at this step:

1. Image is found on both the compute resource and the registry 
    - The wizard should move forward. Additionally, the wizard should show a green tick next to the search box. I think this use case works already
2. Image is found on the compute resource but NOT on the registry
    - The wizard should move forward and show a green tick. I think this use case works already.
3. Image is NOT found on the compute resource but it's found on the registry
    - The wizard should move forward. The image is automatically downloaded when the container is created. This is NOT working with this PR, it will not let users move to the next step when the image is not in the registry.
4. Image is NOT found on the compute resource and is NOT found on the registry
    - The wizard should not move forward because there is no way to create the container without an image.


An example of number 3, if you do not have the busybox image downloaded in the Docker host:
screenshot from 2017-02-17 08-57-16

File: app/models/docker_container_wizard_states/image.rb (hunk 1 above)
Comment (bastilian) [inline app/models/docker_container_wizard_states/image.rb]: Yes, I missed putting it as a inline help for the field. Gonna do that right away.

**Comment by bastilian** [comment]
At the moment the autocomplete for the image name search only checks via `::Docker::Image.exist?`, which only considers local images, the same approach I took for the validator.

I will address that by extending both to trigger a search on the registry in case it can't find an image on the compute resource itself. 

It would also make sense to then enable the image name search as real autocomplete again, since it will need to trigger an API call to verify and turn the check mark green when it's available.

**Comment by bastilian** [comment]
@dLobatog just pushed some changes, I still need to pick the changes apart into branches, but wanted to have it pushed for others to test the end result. You can a better change history in my working branch: https://github.com/bastilian/foreman-docker/tree/fix/18518

**Comment by dLobatog** [comment]
@bastilian Thanks - will take a look in a few

**Comment by thomasmckay** [comment]
When I use registry.access.redhat.com and image openshift3/ose-docker-registry the tag field gives error:
<pre>
2017-02-27T14:06:27 ed796804 [app] [I] Started GET "/image_search/1/auto_complete?search=openshift3%2Fose-docker-registry%3A&registry_id=1&tags=true" for 127.0.0.1 at 2017-02-27 14:06:27 +0000
2017-02-27T14:06:27 ed796804 [app] [I] Processing by ImageSearchController#auto_complete as JSON
2017-02-27T14:06:27 ed796804 [app] [I]   Parameters: {"search"=>"openshift3/ose-docker-registry:", "registry_id"=>"1", "tags"=>"true", "id"=>"1"}
2017-02-27T14:06:27 ed796804 [app] [D] Setting current user thread-local variable to admin
2017-02-27T14:06:27 ed796804 [app] [D] Setting current organization thread-local variable to Example Corporation
2017-02-27T14:06:27 ed796804 [app] [D] Setting current location thread-local variable to none
2017-02-27T14:06:27 ed796804 [app] [D] <Array> [:get, "/v1/repositories/openshift3/ose-docker-registry/tags", {}, nil]
2017-02-27T14:06:28 ed796804 [app] [I] Completed 500 Internal Server Error in 582ms (ActiveRecord: 3.9ms)
2017-02-27T14:06:28 ed796804 [app] [F] 
 | TypeError (no implicit conversion of String into Integer):
 |   /home/vagrant/code/foreman-docker/app/services/foreman_docker/image_search.rb:59:in `[]'
 |   /home/vagrant/code/foreman-docker/app/services/foreman_docker/image_search.rb:59:in `block in registry_tags'
 |   /home/vagrant/code/foreman-docker/app/services/foreman_docker/image_search.rb:59:in `each'
 |   /home/vagrant/code/foreman-docker/app/services/foreman_docker/image_search.rb:59:in `map'
 |   /home/vagrant/code/foreman-docker/app/services/foreman_docker/image_search.rb:59:in `registry_tags'
 |   /home/vagrant/code/foreman-docker/app/services/foreman_docker/image_search.rb:30:in `tags'
 |   /home/vagrant/code/foreman-docker/app/services/foreman_docker/image_search.rb:16:in `search'
 |   app/controllers/concerns/application_shared.rb:14:in `set_timezone'
 |   app/models/concerns/foreman/thread_session.rb:32:in `clear_thread'
 |   lib/middleware/catch_json_parse_errors.rb:8:in `call'
 |   lib/middleware/tagged_logging.rb:18:in `call'
</pre>

**Comment by thomasmckay** [comment]
Note that not all registries have search so failure to find the image should not block proceeding to launch. Specifically, pulp's crane (the registry used in the foreman plugin katello) does not provide search by default.

**Comment by bastilian** [comment]
@thomasmckay This should be fine, the autocomplete won't work, for image names at least. Tags should autocomplete and the validation should pass as well, as it is only based on querying the tags endpoints. If these registries have the tags endpoints missing or disabled, that would be a problem, but shouldn't be hard to address if we have a different way of checking on these.

We should have an availability check in place that does block to prevent host creation failures.

File: app/views/containers/steps/_image_hub_tab.html.erb [hunk 2]
```diff
@@ -4,50 +4,59 @@
...
+
+  <%= text_f f, :tag,
+    :control_group_id => "#{registry}_tag_search",
+    :'data-registry' => registry,
+    :'data-tag' => true,
+    :'data-min-length' => 0,
+    :'data-url' => auto_complete_image_search_path(model.compute_resource_id),
+    :help_inline => spinner('', class: 'hide autocomplete-status') %>
+
+  <div class="col-md-6 col-md-offset-1">
```
Comment (Rohoover) [inline app/views/containers/steps/_image_hub_tab.html.erb]: @dLobatog I prefer it aligned with the fields vs the label. It's not typical to see buttons aligned with labels in the application. Additionally -- the icons on the tabs - that's also unusual.

Comment (bastilian) [inline app/views/containers/steps/_image_hub_tab.html.erb]: @dLobatog & @Rohoover I'll align the search button to be aligned with the inputs. Should we also align the "Registry" label and select to be one with the others?

Comment (dLobatog) [inline app/views/containers/steps/_image_hub_tab.html.erb]: @bastilian I think so - can the alignments can go in a separate PR (seems like something that doesn't require a lot of code)?

Comment (Rohoover) [inline app/views/containers/steps/_image_hub_tab.html.erb]: @bastilian Yes I would align Registry as well.

**Comment by bastilian** [comment]
Closing in favour of #186
